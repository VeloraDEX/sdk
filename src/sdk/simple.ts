import { constructPartialSDK, SDKConfig } from './partial';
import {
  GetAdaptersFunctions,
  constructGetAdapters,
} from '../methods/swap/adapters';
import {
  GetBalancesFunctions,
  constructGetBalances,
} from '../methods/swap/balance';
import { GetRateFunctions, constructGetRate } from '../methods/swap/rates';
import {
  GetSpenderFunctions,
  constructGetSpender,
} from '../methods/swap/spender';
import { GetTokensFunctions, constructGetTokens } from '../methods/swap/token';
import {
  BuildTxFunctions,
  constructBuildTx,
} from '../methods/swap/transaction';
import { constructSwapTx, GetSwapTxFunctions } from '../methods/swap/swapTx';

import {
  constructAxiosFetcher,
  constructFetchFetcher,
  constructEthersV5ContractCaller,
  constructEthersV6ContractCaller,
  constructWeb3ContractCaller,
  constructViemContractCaller,
} from '../helpers';

import type {
  ConstructBaseInput,
  ConstructFetchInput,
  ContractCallerFunctions,
  TransactionContractCallerFn,
  TxHash,
  Address,
  FetcherFunction,
  ExtraFetchParams,
} from '../types';

import type { EthersProviderDeps, MinViemClient } from '../helpers';
import type Web3 from 'web3';

import type { SwapSDKMethods } from '../methods/swap';
import {
  BuildLimitOrderFunctions,
  constructBuildLimitOrder,
} from '../methods/limitOrders/buildOrder';
import {
  constructPostLimitOrder,
  PostLimitOrderFunctions,
} from '../methods/limitOrders/postOrder';
import {
  constructGetLimitOrders,
  GetLimitOrdersFunctions,
} from '../methods/limitOrders/getOrders';
import {
  constructGetLimitOrdersContract,
  GetLimitOrdersContractFunctions,
} from '../methods/limitOrders/getOrdersContract';
import {
  constructBuildLimitOrderTx,
  BuildLimitOrdersTxFunctions,
} from '../methods/limitOrders/transaction';
import {
  constructAllLimitOrdersHandlers,
  LimitOrderHandlers,
} from '../methods/limitOrders';

import {
  constructGetNFTOrdersContract,
  GetNFTOrdersContractFunctions,
} from '../methods/nftOrders/getOrdersContract';
import {
  constructGetNFTOrders,
  GetNFTOrdersFunctions,
} from '../methods/nftOrders/getOrders';
import {
  BuildNFTOrderFunctions,
  constructBuildNFTOrder,
} from '../methods/nftOrders/buildOrder';
import {
  constructPostNFTOrder,
  PostNFTOrderFunctions,
} from '../methods/nftOrders/postOrder';
import {
  constructBuildNFTOrderTx,
  BuildNFTOrdersTxFunctions,
} from '../methods/nftOrders/transaction';
import {
  constructAllNFTOrdersHandlers,
  NFTOrderHandlers,
} from '../methods/nftOrders';

import { constructSwapSDK } from '../methods/swap';
import type { AxiosRequirement } from '../helpers/fetchers/axios';
import { API_URL, DEFAULT_VERSION } from '../constants';
import {
  constructAllDeltaOrdersHandlers,
  DeltaOrderHandlers,
} from '../methods/delta';
import {
  BuildDeltaOrderFunctions,
  constructBuildDeltaOrder,
} from '../methods/delta/buildDeltaOrder';
import {
  constructGetDeltaOrders,
  GetDeltaOrdersFunctions,
} from '../methods/delta/getDeltaOrders';
import {
  constructGetDeltaPrice,
  GetDeltaPriceFunctions,
} from '../methods/delta/getDeltaPrice';
import {
  constructGetDeltaContract,
  GetDeltaContractFunctions,
} from '../methods/delta/getDeltaContract';
import {
  constructGetPartnerFee,
  GetPartnerFeeFunctions,
} from '../methods/delta/getPartnerFee';
import {
  constructPostDeltaOrder,
  PostDeltaOrderFunctions,
} from '../methods/delta/postDeltaOrder';
import {
  constructGetQuote,
  GetQuoteFunctions,
} from '../methods/quote/getQuote';
import {
  constructGetBridgeInfo,
  GetBridgeInfoFunctions,
} from '../methods/delta/getBridgeInfo';
import {
  constructGetMulticallHandlers,
  GetMulticallHandlersFunctions,
} from '../methods/delta/getMulticallHandlers';
import {
  constructIsTokenSupportedInDelta,
  IsTokenSupportedInDeltaFunctions,
} from '../methods/delta/isTokenSupportedInDelta';
import {
  BuildCrosschainOrderBridgeFunctions,
  constructBuildCrosschainOrderBridge,
} from '../methods/delta/buildCrosschainOrderBridge';

export type SwapFetchMethods = GetBalancesFunctions &
  GetTokensFunctions &
  GetSpenderFunctions &
  BuildTxFunctions &
  GetAdaptersFunctions &
  GetRateFunctions &
  GetSwapTxFunctions;

export type LimitOrdersFetchMethods = GetLimitOrdersContractFunctions &
  GetLimitOrdersFunctions &
  BuildLimitOrderFunctions &
  PostLimitOrderFunctions &
  BuildLimitOrdersTxFunctions;

export type NFTOrdersFetchMethods = GetNFTOrdersContractFunctions &
  GetNFTOrdersFunctions &
  BuildNFTOrderFunctions &
  PostNFTOrderFunctions &
  BuildNFTOrdersTxFunctions;

export type DeltaFetchMethods = BuildDeltaOrderFunctions &
  BuildCrosschainOrderBridgeFunctions &
  GetDeltaOrdersFunctions &
  GetDeltaPriceFunctions &
  GetDeltaContractFunctions &
  GetPartnerFeeFunctions &
  GetMulticallHandlersFunctions &
  GetBridgeInfoFunctions &
  IsTokenSupportedInDeltaFunctions &
  PostDeltaOrderFunctions;

export type SimpleFetchSDK = {
  swap: SwapFetchMethods;
  limitOrders: LimitOrdersFetchMethods;
  nftOrders: NFTOrdersFetchMethods;
  delta: DeltaFetchMethods;
  quote: QuoteFetchMethods;
} & Required<ConstructBaseInput>;

export type QuoteFetchMethods = GetQuoteFunctions;

export type SimpleSDK = {
  swap: SwapSDKMethods<TxHash>;
  limitOrders: LimitOrderHandlers<TxHash>;
  nftOrders: NFTOrderHandlers<TxHash>;
  delta: DeltaOrderHandlers<TxHash>;
  quote: QuoteFetchMethods;
} & Required<ConstructBaseInput>;

export type FetcherOptions = (
  | {
      axios: AxiosRequirement;
    }
  | { fetch: typeof fetch }
  | { fetcher: FetcherFunction }
) &
  ExtraFetchParams;

type SimpleOptions = ConstructBaseInput & FetcherOptions;

export type ProviderOptions = (
  | EthersProviderDeps
  | { web3: Web3 }
  | { viemClient: MinViemClient }
) & {
  account: Address;
};

const constructFetcher = (options: FetcherOptions): FetcherFunction => {
  if ('axios' in options) {
    return constructAxiosFetcher(options.axios, options);
  }
  if ('fetch' in options) {
    return constructFetchFetcher(options.fetch, options);
  }
  return (params) => {
    // adding apiKey to headers if it's provided
    const headers = options?.apiKey
      ? {
          'X-API-KEY': options.apiKey,
          ...params.headers,
          ...params.requestParams?.headers,
        }
      : params.headers;

    return options.fetcher({ ...params, headers });
  };
};

/** @description construct SDK with methods that fetch from API and optionally with blockchain provider calling methods */
export function constructSimpleSDK(options: SimpleOptions): SimpleFetchSDK;

export function constructSimpleSDK(
  options: SimpleOptions,
  providerOptions: ProviderOptions
): SimpleSDK;
export function constructSimpleSDK(
  options: SimpleOptions,
  providerOptions?: ProviderOptions
): SimpleFetchSDK | SimpleSDK {
  const fetcher = constructFetcher(options);

  if (!providerOptions) {
    const config: ConstructFetchInput = {
      apiURL: options.apiURL,
      chainId: options.chainId,
      version: options.version,
      fetcher,
    };

    // include all available functions that don't need `contractCaller`
    const swap: SwapFetchMethods = constructPartialSDK(
      config,
      constructGetBalances,
      constructGetTokens,
      constructGetSpender,
      constructBuildTx,
      constructGetAdapters,
      constructGetRate,
      constructSwapTx
    );

    const limitOrders = constructPartialSDK(
      config,
      constructBuildLimitOrder,
      constructPostLimitOrder,
      constructGetLimitOrders,
      constructGetLimitOrdersContract,
      constructBuildLimitOrderTx
    );

    const nftOrders = constructPartialSDK(
      config,
      constructBuildNFTOrder,
      constructPostNFTOrder,
      constructGetNFTOrders,
      constructGetNFTOrdersContract,
      constructBuildNFTOrderTx
    );

    const delta = constructPartialSDK(
      config,
      constructBuildDeltaOrder,
      constructBuildCrosschainOrderBridge,
      constructPostDeltaOrder,
      constructGetDeltaOrders,
      constructGetDeltaPrice,
      constructGetDeltaContract,
      constructGetPartnerFee,
      constructGetMulticallHandlers,
      constructGetBridgeInfo,
      constructIsTokenSupportedInDelta
    );

    const quote = constructPartialSDK(config, constructGetQuote);

    return {
      swap,
      limitOrders,
      nftOrders,
      delta,
      quote,
      apiURL: options.apiURL ?? API_URL,
      chainId: options.chainId,
      version: options.version ?? DEFAULT_VERSION,
    };
  }

  const contractCaller = constructSimpleContractCaller(providerOptions);

  const config: SDKConfig<TxHash> = {
    apiURL: options.apiURL,
    version: options.version,
    chainId: options.chainId,
    fetcher,
    contractCaller,
  };

  const swap: SwapSDKMethods<TxHash> = constructSwapSDK(config);

  const limitOrders: LimitOrderHandlers<TxHash> =
    constructAllLimitOrdersHandlers<TxHash>(config);

  const nftOrders: NFTOrderHandlers<TxHash> =
    constructAllNFTOrdersHandlers<TxHash>(config);

  const delta: DeltaOrderHandlers<TxHash> =
    constructAllDeltaOrdersHandlers<TxHash>(config);

  const quote = constructGetQuote(config);

  return {
    swap,
    limitOrders,
    nftOrders,
    delta,
    quote,
    apiURL: options.apiURL ?? API_URL,
    chainId: options.chainId,
    version: options.version ?? DEFAULT_VERSION,
  };
}

function constructSimpleContractCaller(
  providerOptions: ProviderOptions
): ContractCallerFunctions<TxHash> {
  if ('ethersProviderOrSigner' in providerOptions) {
    const {
      staticCall,
      transactCall: _transactCall,
      signTypedDataCall,
    } = constructEthersV5ContractCaller(
      providerOptions,
      providerOptions.account
    );

    const transactCall: TransactionContractCallerFn<TxHash> = async (
      params
    ) => {
      const contractTx = await _transactCall(params);

      // as soon as tx is sent
      // returning tx hash, it's up to the user to wait for tx
      return contractTx.hash as TxHash;
    };

    return { staticCall, transactCall, signTypedDataCall };
  }

  if ('ethersV6ProviderOrSigner' in providerOptions) {
    const {
      staticCall,
      transactCall: _transactCall,
      signTypedDataCall,
    } = constructEthersV6ContractCaller(
      providerOptions,
      providerOptions.account
    );

    const transactCall: TransactionContractCallerFn<TxHash> = async (
      params
    ) => {
      const contractTx = await _transactCall(params);

      // as soon as tx is sent
      // returning tx hash, it's up to the user to wait for tx
      return contractTx.hash as TxHash;
    };

    return { staticCall, transactCall, signTypedDataCall };
  }

  if ('viemClient' in providerOptions) {
    const contractCaller = constructViemContractCaller(
      providerOptions.viemClient,
      providerOptions.account
    );
    return contractCaller;
  }

  const {
    staticCall,
    transactCall: _transactCall,
    signTypedDataCall,
  } = constructWeb3ContractCaller(
    providerOptions.web3,
    providerOptions.account
  );

  const transactCall: TransactionContractCallerFn<TxHash> = async (params) => {
    const unpromiEvent = await _transactCall(params);

    // as soon as tx is sent
    // returning tx hash, it's up to the user to wait for tx
    return new Promise<TxHash>((resolve, reject) => {
      unpromiEvent.once('transactionHash', (hash) => resolve(hash as TxHash));
      unpromiEvent.once('error', reject);
    });
  };

  return { staticCall, transactCall, signTypedDataCall };
}
