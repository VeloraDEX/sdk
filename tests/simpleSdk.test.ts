import * as dotenv from 'dotenv';
import Web3 from 'web3';
import { BigNumber as BigNumberEthers, ethers } from 'ethersV5';
import { ethers as ethersV6 } from 'ethers';
import axios from 'axios';
import fetch from 'isomorphic-unfetch';
import { isAllowance, SwapSide, SimpleFetchSDK } from '../src';
import BigNumber from 'bignumber.js';

import erc20abi from './abi/ERC20.json';

import { assert } from 'ts-essentials';

import { constructSimpleSDK, SimpleSDK } from '../src/sdk/simple';
import { HardhatProvider, setupFork } from './helpers/hardhat';

dotenv.config();

jest.setTimeout(30 * 1000);

const ETH = '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee';
const DAI = '0x6B175474E89094C44Da98b954EedeAC495271d0F';
// const HEX = '0x2b591e99afe9f32eaa6214f7b7629768c40eeb39';

const DUMMY_ADDRESS_FOR_TESTING_ALLOWANCES =
  '0xb9A079479A7b0F4E7F398F7ED3946bE6d9a40E79';

const chainId = 1;
const srcToken = ETH;
const destToken = DAI;
const srcAmount = (1 * 1e18).toString(); //The source amount multiplied by its decimals

const referrer = 'sdk-test';
const TEST_MNEMONIC =
  // fresh throw-away mnemonic, acc from previous public mnemonic was overused and generated errors on BUY
  'venue olympic vendor sign era tongue sleep peace daughter enjoy dry august';
//0xbB12cA3d135e40517CDDAc23E9fBE0E66f53e829
const wallet = ethers.Wallet.fromMnemonic(TEST_MNEMONIC);
const walletV6 = ethersV6.HDNodeWallet.fromPhrase(TEST_MNEMONIC);

const web3provider = new Web3(HardhatProvider as any);

const ethersProvider = new ethers.providers.Web3Provider(
  HardhatProvider as any
);
const ethersV6Provider = new ethersV6.BrowserProvider(HardhatProvider);
const signerV6 = walletV6.connect(ethersV6Provider);

const signer = wallet.connect(ethersProvider);
const senderAddress = signer.address;

describe.each([
  ['fetch', { fetch }],
  ['axios', { axios }],
])('Simple SDK: fetcher made with: %s', (testName, fetcherOptions) => {
  let sdk: SimpleFetchSDK;

  beforeAll(async () => {
    await setupFork({ accounts: [{ address: senderAddress, balance: 8e18 }] });

    sdk = constructSimpleSDK({ chainId, ...fetcherOptions, version: '6.2' });
  });
  test('getBalance', async () => {
    try {
      const balance = await sdk.swap.getBalance(senderAddress, ETH);
      expect(balance).toBeDefined();
    } catch (error: any) {
      // workaround for API sometimes failing on some Tokens(?)
      expect(error.message).toMatch(/Only chainId \d+ is supported/);
    }
  });

  test('Get_Markets', async () => {
    const markets = await sdk.swap.getAdapters();
    expect(markets.length).toBeGreaterThan(15);
  });

  test('Get_Tokens', async () => {
    const tokens = await sdk.swap.getTokens();

    expect(Array.isArray(tokens)).toBe(true);
    expect(tokens.length).toBeGreaterThan(0);
    expect(tokens[0]).toEqual(
      expect.objectContaining({
        symbol: expect.any(String),
        address: expect.any(String),
        decimals: expect.any(Number),
      })
    );
  });

  test('Get_Rates', async () => {
    const priceRoute = await sdk.swap.getRate({
      srcToken: ETH,
      destToken: DAI,
      amount: srcAmount,
      userAddress: senderAddress,
      side: SwapSide.SELL,
      options: {
        includeDEXS: ['UniswapV2'],
        otherExchangePrices: true,
      },
    });

    const { destAmount, bestRoute, others } = priceRoute;

    expect(typeof destAmount).toBe('string');

    expect(Array.isArray(bestRoute)).toBe(true);

    const swapExchange = bestRoute[0]?.swaps[0]?.swapExchanges[0];

    assert(swapExchange, 'exchange available at swapExchanges[0]');

    expect(typeof swapExchange.destAmount).toBe('string');
    expect(new BigNumber(swapExchange.destAmount).isNaN()).toBe(false);

    expect(typeof swapExchange.exchange).toBe('string');

    const firstBestRoute = bestRoute[0];
    assert(firstBestRoute, 'route available at bestRoute[0]');

    expect(typeof firstBestRoute.percent).toBe('number');
    expect(new BigNumber(firstBestRoute.percent).isNaN()).toBe(false);

    expect(typeof swapExchange.srcAmount).toBe('string');
    expect(new BigNumber(swapExchange.srcAmount).isNaN()).toBe(false);

    expect(Array.isArray(others)).toBe(true);

    const firstRoute = others?.[0];

    assert(firstRoute, 'at least one route must exist');

    expect(typeof firstRoute.exchange).toBe('string');

    expect(typeof firstRoute.unit).toBe('string');
    expect(firstRoute.unit && new BigNumber(firstRoute.unit).isNaN()).toBe(
      false
    );
  });

  test('Get_SwapTxData', async () => {
    const { priceRoute, txParams } = await sdk.swap.getSwapTxData({
      srcToken: ETH,
      destToken: DAI,
      amount: srcAmount,
      userAddress: senderAddress,
      side: SwapSide.SELL,
      slippage: 500,
      options: {
        includeDEXS: ['UniswapV2'],
        includeContractMethods: ['swapExactAmountIn'], // more stable results
      },
    });

    const bestRouteStable = priceRoute.bestRoute.map((b) => ({
      ...b,
      swaps: b.swaps.map((s) => ({
        ...s,
        //                             grab first only
        swapExchanges: s.swapExchanges.slice(0, 1).map((se) => ({
          ...se,
          // numbers change based on how many exchanges are in the route
          srcAmount: 'dynamic_number',
          destAmount: 'dynamic_number',
          data: 'dynamic_object',
          percent: 'dynamic_number',
          poolAddresses: 'dynamic_array',
          poolIdentifiers: 'dynamic_array',
        })),
      })),
    }));

    const priceRouteStable = {
      ...priceRoute,
      gasCost: 'dynamic_number',
      gasCostUSD: 'dynamic_number',
      hmac: 'dynamic_string',
      destAmount: 'dynamic_number',
      blockNumber: 'dynamic_number',
      srcUSD: 'dynamic_number',
      destUSD: 'dynamic_number',
      bestRoute: bestRouteStable,
    };

    const txParamsStable = {
      ...txParams,
      data: 'dynamic_string',
      from: 'dynamic_string',
      gasPrice: 'dynamic_number',
    };

    expect(txParams.from).toEqual(senderAddress);

    expect(priceRouteStable).toMatchSnapshot('Get_SwapTxData::priceRoute');
    expect(txParamsStable).toMatchSnapshot('Get_SwapTxData::txParams');
  });

  test('Get_Spender', async () => {
    const spender = await sdk.swap.getSpender();
    expect(web3provider.utils.isAddress(spender));
  });

  test('Get_Allowance', async () => {
    const allowance = await sdk.swap.getAllowance(
      DUMMY_ADDRESS_FOR_TESTING_ALLOWANCES,
      DAI
    );

    assert(isAllowance(allowance), 'hardcoded dummy address should be found');

    expect(allowance.allowance).toEqual('123000000000000000');
  });

  test('Get_Adapters', async () => {
    const adapters = await sdk.swap.getAdapters();
    expect(adapters).toMatchSnapshot('Get_Adapters');
  });

  test('Build_Tx', async () => {
    const destToken = '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48';
    const priceRoute = await sdk.swap.getRate({
      srcToken,
      destToken,
      amount: srcAmount,
      userAddress: senderAddress,
      side: SwapSide.SELL,
      options: {
        includeDEXS: ['UniswapV2'],
      },
    });

    const destAmount = new BigNumber(priceRoute.destAmount)
      .times(0.99)
      .toFixed(0);

    const txParams = await sdk.swap.buildTx(
      {
        srcToken,
        destToken,
        srcAmount,
        destAmount,
        priceRoute,
        userAddress: senderAddress,
        partner: referrer,
      },
      { ignoreChecks: true }
    );

    expect(typeof txParams).toBe('object');
  });
  test('Build_and_Send_Tx', async () => {
    const priceRoute = await sdk.swap.getRate({
      srcToken,
      destToken,
      amount: srcAmount,
      userAddress: senderAddress,
      side: SwapSide.SELL,
      options: {
        includeDEXS: ['Uniswap', 'UniswapV2', 'Balancer', 'Oasis'],
      },
    });

    const destAmount = new BigNumber(priceRoute.destAmount)
      .times(0.99)
      .toFixed(0);

    const txParams = await sdk.swap.buildTx(
      {
        srcToken,
        destToken,
        srcAmount,
        destAmount,
        priceRoute,
        userAddress: signer.address,
        partner: referrer,
      },
      { ignoreChecks: true }
    );

    const transaction = {
      ...txParams,
      gasPrice:
        txParams.gasPrice &&
        '0x' + new BigNumber(txParams.gasPrice).toString(16),
      maxFeePerGas:
        txParams.maxFeePerGas &&
        '0x' + new BigNumber(txParams.maxFeePerGas).toString(16),
      maxPriorityFeePerGas:
        txParams.maxPriorityFeePerGas &&
        '0x' + new BigNumber(txParams.maxPriorityFeePerGas).toString(16),
      gasLimit: '0x' + new BigNumber(5000000).toString(16),
      value: '0x' + new BigNumber(txParams.value).toString(16),
    };
    const toContract = new ethers.Contract(destToken, erc20abi, ethersProvider);
    const beforeFromBalance = await ethersProvider.getBalance(signer.address);
    const beforeToBalance = await toContract.balanceOf(signer.address);

    const txr = await signer.sendTransaction(transaction);
    await txr.wait(1);
    const afterFromBalance = await ethersProvider.getBalance(signer.address);
    const afterToBalance = await toContract.balanceOf(signer.address);
    expect(beforeFromBalance.gt(afterFromBalance)).toBeTruthy();
    expect(beforeToBalance.lt(afterToBalance)).toBeTruthy();
  });
  test('Build_and_Send_Tx_BUY', async () => {
    const destAmount = srcAmount;
    const priceRoute = await sdk.swap.getRate({
      srcToken,
      destToken,
      amount: destAmount,
      userAddress: senderAddress,
      side: SwapSide.BUY,
    });
    const _srcAmount = new BigNumber(priceRoute.srcAmount)
      .times(1.1)
      .toFixed(0);

    const txParams = await sdk.swap.buildTx(
      {
        srcToken,
        destToken,
        srcAmount: _srcAmount,
        destAmount,
        priceRoute,
        userAddress: signer.address,
        partner: referrer,
      },
      { ignoreChecks: true }
    );

    const transaction = {
      ...txParams,
      gasPrice:
        txParams.gasPrice &&
        '0x' + new BigNumber(txParams.gasPrice).toString(16),
      maxFeePerGas:
        txParams.maxFeePerGas &&
        '0x' + new BigNumber(txParams.maxFeePerGas).toString(16),
      maxPriorityFeePerGas:
        txParams.maxPriorityFeePerGas &&
        '0x' + new BigNumber(txParams.maxPriorityFeePerGas).toString(16),
      gasLimit: '0x' + new BigNumber(5000000).toString(16),
      value: '0x' + new BigNumber(txParams.value).toString(16),
    };
    const toContract = new ethers.Contract(destToken, erc20abi, ethersProvider);
    const beforeFromBalance = await ethersProvider.getBalance(signer.address);
    const beforeToBalance = await toContract.balanceOf(signer.address);

    const txr = await signer.sendTransaction(transaction);
    await txr.wait(1);
    const afterFromBalance = await ethersProvider.getBalance(signer.address);
    const afterToBalance = await toContract.balanceOf(signer.address);
    expect(beforeFromBalance.gt(afterFromBalance)).toBeTruthy();
    expect(beforeToBalance.lt(afterToBalance)).toBeTruthy();
  });
});

describe.each([
  [
    'fetch & ethersV5',
    { fetch },
    {
      ethersProviderOrSigner: signer,
      EthersContract: ethers.Contract,
      account: senderAddress,
    },
  ],
  [
    'fetch & ethersV6',
    { fetch },
    {
      ethersV6ProviderOrSigner: signerV6,
      EthersV6Contract: ethersV6.Contract,
      account: senderAddress,
    },
  ],
  ['axios & web3', { axios }, { web3: web3provider, account: senderAddress }],
])(
  'Simple SDK: contract calling methods: %s',
  (testName, fetcherOptions, providerOptions) => {
    let sdk: SimpleSDK;

    beforeAll(() => {
      sdk = constructSimpleSDK(
        { chainId, ...fetcherOptions, version: '5' },
        providerOptions
      );
    });
    test('approveToken', async () => {
      const txHash = await sdk.swap.approveToken('12345', DAI);

      await ethersProvider.waitForTransaction(txHash);

      const toContract = new ethers.Contract(
        destToken,
        erc20abi,
        ethersProvider
      );
      const spender = await sdk.swap.getSpender();
      const allowance: BigNumberEthers = await toContract.allowance(
        signer.address,
        spender
      );
      expect(allowance.toString()).toEqual('12345');
    });
  }
);
