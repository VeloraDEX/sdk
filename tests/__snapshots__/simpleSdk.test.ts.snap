// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple SDK: fetcher made with: axios Get_Adapters: Get_Adapters 1`] = `
[
  "Bancor",
  "Compound",
  "DODOV1",
  "DODOV2",
  "TraderJoeV2.1",
  "Smoothy",
  "CurveV2",
  "Lido",
  "EtherFi",
  "Stader",
  "Bebop",
  "CurveV1",
  "Swerve",
  "BalancerV1",
  "BalancerV2",
  "BalancerV3",
  "UniswapV2",
  "DefiSwap",
  "LinkSwap",
  "SushiSwap",
  "ShibaSwap",
  "Verse",
  "UniswapV3",
  "SushiSwapV3",
  "UniswapV4",
  "PancakeSwapV2",
  "PancakeswapV3",
  "AaveV2",
  "AaveV3",
  "AaveV3Lido",
  "KyberDmm",
  "Weth",
  "PolygonMigrator",
  "Synapse",
  "ParaSwapLimitOrders",
  "SolidlyV2",
  "Synthetix",
  "CurveV1Factory",
  "CurveV1StableNg",
  "wstETH",
  "wUSDM",
  "wUSDL",
  "sUSDe",
  "stcUSD",
  "Hashflow",
  "MaverickV1",
  "MaverickV2",
  "SwaapV2",
  "AngleTransmuter",
  "AngleStakedStableUSD",
  "AngleStakedStableEUR",
  "SolidlyV3",
  "Wombat",
  "Swell",
  "Spark",
  "sUSDS",
  "AaveV3Stata",
  "AaveV3StataV2",
  "OSwap",
  "ConcentratorArusd",
  "FxProtocolRusd",
  "AaveGsm",
  "LitePsm",
  "UsualBond",
  "StkGHO",
  "DaiUsds",
  "MkrSky",
  "FluidDex",
  "FluidDexLite",
  "UsdcUsualUSDC",
  "UsualUSDCUsd0",
  "UsualMWrappedM",
  "MWrappedM",
  "WrappedMM",
  "UsualMUsd0",
  "UsualPP",
  "Ekubo",
  "MiroMigrator",
  "AaveV3Pendle",
  "RingV2",
  "BunniV2",
  "Cap",
  "AugustusRFQ",
]
`;

exports[`Simple SDK: fetcher made with: axios Get_SwapTxData: Get_SwapTxData::priceRoute 1`] = `
{
  "bestRoute": [
    {
      "percent": 100,
      "swaps": [
        {
          "destDecimals": 18,
          "destToken": "0x6b175474e89094c44da98b954eedeac495271d0f",
          "srcDecimals": 18,
          "srcToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
          "swapExchanges": [
            {
              "data": "dynamic_object",
              "destAmount": "dynamic_number",
              "exchange": "UniswapV2",
              "percent": "dynamic_number",
              "poolAddresses": "dynamic_array",
              "poolIdentifiers": "dynamic_array",
              "srcAmount": "dynamic_number",
            },
          ],
        },
      ],
    },
  ],
  "blockNumber": "dynamic_number",
  "contractAddress": "0x6a000f20005980200259b80c5102003040001068",
  "contractMethod": "swapExactAmountIn",
  "destAmount": "dynamic_number",
  "destDecimals": 18,
  "destToken": "0x6b175474e89094c44da98b954eedeac495271d0f",
  "destUSD": "dynamic_number",
  "gasCost": "dynamic_number",
  "gasCostUSD": "dynamic_number",
  "hmac": "dynamic_string",
  "maxImpactReached": false,
  "network": 1,
  "partner": "anon",
  "partnerFee": 0,
  "side": "SELL",
  "srcAmount": "1000000000000000000",
  "srcDecimals": 18,
  "srcToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
  "srcUSD": "dynamic_number",
  "tokenTransferProxy": "0x6a000f20005980200259b80c5102003040001068",
  "version": "6.2",
}
`;

exports[`Simple SDK: fetcher made with: axios Get_SwapTxData: Get_SwapTxData::txParams 1`] = `
{
  "chainId": 1,
  "data": "dynamic_string",
  "from": "dynamic_string",
  "gasPrice": "dynamic_number",
  "to": "0x6a000f20005980200259b80c5102003040001068",
  "value": "1000000000000000000",
}
`;

exports[`Simple SDK: fetcher made with: fetch Get_Adapters: Get_Adapters 1`] = `
[
  "Bancor",
  "Compound",
  "DODOV1",
  "DODOV2",
  "TraderJoeV2.1",
  "Smoothy",
  "CurveV2",
  "Lido",
  "EtherFi",
  "Stader",
  "Bebop",
  "CurveV1",
  "Swerve",
  "BalancerV1",
  "BalancerV2",
  "BalancerV3",
  "UniswapV2",
  "DefiSwap",
  "LinkSwap",
  "SushiSwap",
  "ShibaSwap",
  "Verse",
  "UniswapV3",
  "SushiSwapV3",
  "UniswapV4",
  "PancakeSwapV2",
  "PancakeswapV3",
  "AaveV2",
  "AaveV3",
  "AaveV3Lido",
  "KyberDmm",
  "Weth",
  "PolygonMigrator",
  "Synapse",
  "ParaSwapLimitOrders",
  "SolidlyV2",
  "Synthetix",
  "CurveV1Factory",
  "CurveV1StableNg",
  "wstETH",
  "wUSDM",
  "wUSDL",
  "sUSDe",
  "stcUSD",
  "Hashflow",
  "MaverickV1",
  "MaverickV2",
  "SwaapV2",
  "AngleTransmuter",
  "AngleStakedStableUSD",
  "AngleStakedStableEUR",
  "SolidlyV3",
  "Wombat",
  "Swell",
  "Spark",
  "sUSDS",
  "AaveV3Stata",
  "AaveV3StataV2",
  "OSwap",
  "ConcentratorArusd",
  "FxProtocolRusd",
  "AaveGsm",
  "LitePsm",
  "UsualBond",
  "StkGHO",
  "DaiUsds",
  "MkrSky",
  "FluidDex",
  "FluidDexLite",
  "UsdcUsualUSDC",
  "UsualUSDCUsd0",
  "UsualMWrappedM",
  "MWrappedM",
  "WrappedMM",
  "UsualMUsd0",
  "UsualPP",
  "Ekubo",
  "MiroMigrator",
  "AaveV3Pendle",
  "RingV2",
  "BunniV2",
  "Cap",
  "AugustusRFQ",
]
`;

exports[`Simple SDK: fetcher made with: fetch Get_SwapTxData: Get_SwapTxData::priceRoute 1`] = `
{
  "bestRoute": [
    {
      "percent": 100,
      "swaps": [
        {
          "destDecimals": 18,
          "destToken": "0x6b175474e89094c44da98b954eedeac495271d0f",
          "srcDecimals": 18,
          "srcToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
          "swapExchanges": [
            {
              "data": "dynamic_object",
              "destAmount": "dynamic_number",
              "exchange": "UniswapV2",
              "percent": "dynamic_number",
              "poolAddresses": "dynamic_array",
              "poolIdentifiers": "dynamic_array",
              "srcAmount": "dynamic_number",
            },
          ],
        },
      ],
    },
  ],
  "blockNumber": "dynamic_number",
  "contractAddress": "0x6a000f20005980200259b80c5102003040001068",
  "contractMethod": "swapExactAmountIn",
  "destAmount": "dynamic_number",
  "destDecimals": 18,
  "destToken": "0x6b175474e89094c44da98b954eedeac495271d0f",
  "destUSD": "dynamic_number",
  "gasCost": "dynamic_number",
  "gasCostUSD": "dynamic_number",
  "hmac": "dynamic_string",
  "maxImpactReached": false,
  "network": 1,
  "partner": "anon",
  "partnerFee": 0,
  "side": "SELL",
  "srcAmount": "1000000000000000000",
  "srcDecimals": 18,
  "srcToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
  "srcUSD": "dynamic_number",
  "tokenTransferProxy": "0x6a000f20005980200259b80c5102003040001068",
  "version": "6.2",
}
`;

exports[`Simple SDK: fetcher made with: fetch Get_SwapTxData: Get_SwapTxData::txParams 1`] = `
{
  "chainId": 1,
  "data": "dynamic_string",
  "from": "dynamic_string",
  "gasPrice": "dynamic_number",
  "to": "0x6a000f20005980200259b80c5102003040001068",
  "value": "1000000000000000000",
}
`;
